netket/variational_states/base.py

```diff
@@ -0,0 +1,151 @@
+import abc
+
+from functools import partial
+from typing import Any, Optional, Tuple
+
+import flax
+
+from netket.operator import AbstractOperator, LocalLiouvillian
+from netket.hilbert import AbstractHilbert, DoubledHilbert
+from netket.stats import Stats
+
+PyTree = Any
+
+
+class VariationalState(abc.ABC):
+    """Abstract class for variational states representing either pure states
+    or mixed quantum states.
+    A variational state is a quantum state depending on a set of
+    parameters, and that supports operations such
+    as computing quantum expectation values and their gradients."""
+
+    _hilbert: AbstractHilbert
+    """The hilbert space on which this state is defined."""
+
+    parameters: PyTree
+    """The PyTree of variational parameters"""
+
+    def __init__(self, hilbert):
+        self._hilbert = hilbert
+
+        self._model_state = {}
+        self._parameters = {}
+
+    @property
+    def hilbert(self) -> AbstractHilbert:
+        r"""netket.hilbert.AbstractHilbert: The descriptor of the Hilbert space
+        on which this variational state is defined.
+        """
+        return self._hilbert
+
+    @property
+    def parameters(self) -> PyTree:
+        return self._parameters
+
+    @parameters.setter
+    def parameters(self, pars: PyTree):
+        self._parameters = pars
+
+    @property
+    def model_state(self) -> PyTree:
+        return self._model_state
+
+    @model_state.setter
+    def model_state(self, state: PyTree):
+        self._model_state = state
+
+    @property
+    def variables(self) -> PyTree:
+        return flax.core.freeze({"params": self.parameters, **self.model_state})
+
+    @variables.setter
+    def variables(self, vars) -> PyTree:
+        self.model_state, self.parameters = vars.pop("params")
```

Comment:
If a user tries to pass  his own `apply_fun` and `variables` to e.g. `nk.variational_states.ClassicalVariationalState`, then this only works for  `FrozenDict`, but not if a normal python dict / pytree is passed.
I think in the latter case it should at least give an error or freeze it internally?

Reply:
what do you think of 
```
        if not isinstance(vars, FrozenDict):
            vars = flax.core.freeze(vars)
```